Adds CLP API to Schema router

flovilmart opened the pull request.

drew-gross commented:
Personally I'd rather have this be part of the existing route rather than a new route (ie just add more info to the existing response to `GET /schemas`.

flovilmart commented:
fine by me, what parameter do you suggest 
{
   className: "className",
   permissions: {}
}
it permissions is sent as an empty obj, that locks the object down.

drew-gross commented:
I like long names e.g. `classLevelPermissions`. That aslo leaves room for pointer permissions in a separate key called `pointerPermissions`.

flovilmart commented:
OK good, so on the PUT for the schema, and on the GET, I add the permissions to the current result?

drew-gross commented:
GET sounds right, for PUT I think just replace the permission on the _SCHEMA object entirely with whatever the include in the PUT, and make no modifications if the key isn't included. POST should probably support CLPs too, which shouldn't be too hard. The dashboard won't need it but people using the schemas API for their own purposes would probably find it handy.

flovilmart commented:
alrighty!

drew-gross commented on src/Schema.js:
```diff
@@ -76,6 +76,17 @@ var requiredColumns = {
   _Role: ["name", "ACL"]
 }
 
+let CLPValidKeys = ['find', 'get', 'create', 'update', 'delete'];
```
What about `addField`

flovilmart replied:
Didn't knew it existed :) what's the full list?

flovilmart replied:
btw addField is never checked anywhere.

drew-gross replied:
These 6 are the full list, you can verify using dashboard.parse.com. We should probably start checking addField :p Schema.js should be able to handle it without too much trouble, since there is that 'freeze' parameter.

drew-gross commented on src/Schema.js:
```diff
@@ -76,6 +76,17 @@ var requiredColumns = {
+    return;
+  }
+  Object.keys(perms).forEach((key) => {
+    if (CLPValidKeys.indexOf(key) == -1) {
```
I think this function doesn't go far enough - it should also validate the contents of the keys to make sure they contain only roles and users. Also my preference would be to have it return true/false rather than throw.

drew-gross commented:
I think since this is such a security critical feature we should do much more thorough testing than usual - verify that users/roles don't gain permissions after a failed attempt to change permissions, verify that they don't gain one permission after a different permission was modified, etc. At the very least we need to do some tests for users as well as for roles, which I don't currently see. Other than that and the stricter validation in `validateCLP` this looks good :)

flovilmart commented:
Agreed with all that!

drew-gross commented on spec/schemas.spec.js:
```diff
@@ -872,4 +872,532 @@ describe('schemas', () => {
+    })
+  });
+  
+  it('should not be able to add a field', done => {
```
wrong name

drew-gross commented on spec/schemas.spec.js:
```diff
@@ -872,4 +872,532 @@ describe('schemas', () => {
+    }).then(() => {
+      let query = new Parse.Query('AClass');
+      return query.find().then((err) => {
+        expect(err.message).toEqual('Permission denied for this action.');
```
I think this expect is in the wrong callback

flovilmart replied:
:+1:

drew-gross commented on spec/schemas.spec.js:
```diff
@@ -872,4 +872,532 @@ describe('schemas', () => {
+      fail("should not fail!");
+      done();
+    }).catch( (err) => {
+      console.error(err);
```
No more `console.log` in tests please, they spew too much garbage already.

drew-gross commented on spec/schemas.spec.js:
```diff
@@ -872,4 +872,532 @@ describe('schemas', () => {
+    }).then(() => {
+      let query = new Parse.Query('AClass');
+      return query.find().then((err) => {
+        expect(err.message).toEqual('Permission denied for this action.');
```
Same, expect looks like it's in the wrong callback

drew-gross commented on src/Schema.js:
```diff
@@ -76,6 +76,43 @@ var requiredColumns = {
+    Object.keys(perms[operation]).forEach((key) => {
+      verifyPermissionKey(key);
+      let perm = perms[operation][key];
+      if (perm !== true && perm !== false) {
```
A test or two for permissions explicitly set to false would be nice, I noticed we don't have any.

drew-gross commented:
I see lots of tests for public permissions and for permissions on roles, but still none for users. Also we should have some tests for setting a permission to false explicitly (that really should have been done in the original CLP tests but now is as good a time to add them as any)

flovilmart commented:
@drew-gross I updated, the setting permission to false blacklist, I can revert that if we don't want it now. Added 2 test cases with multiple users, as well as role and public

drew-gross commented on src/Schema.js:
```diff
@@ -776,17 +875,20 @@ function mongoSchemaAPIResponseFields(schema) {
     fields: mongoSchemaAPIResponseFields(schema),
   };
+  if (schema._metadata && schema._metadata.class_permissions) {
+    result.classLevelPermissions = schema._metadata.class_permissions;
```
One last request: Can you have the response always include the classLevelPermissions key? Have it return something equivalent to how the server behaves if there is actually no _metadata or _metadata.class_permissions, which appears to be the same as public everything.

flovilmart replied:
No problem!

drew-gross commented:
One more small request. Sorry for having to go back and forth so many times on this.

flovilmart commented:
Don't be sorry, that's part of the process. In that case should we store it as the default?

drew-gross commented:
We shouldn't be making any database writes as part of a GET query, but I would be cool with saving in POST/PUT if you want to do that. Thats definitely optional though.

drew-gross commented:
Feel free to merge assuming tests pass.
